crates/ruma-federation-api/src/authenticated_media.rs

```diff
@@ -0,0 +1,291 @@
+//! Authenticated endpoints for the content repository, according to [MSC3916].
+//!
+//! [MSC3916]: https://github.com/matrix-org/matrix-spec-proposals/pull/3916
+
+use serde::{Deserialize, Serialize};
+
+pub mod get_content;
+pub mod get_content_thumbnail;
+
+/// The metadata of a file from the content repository.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct ContentMetadata {}
+
+impl ContentMetadata {
+    /// Creates a new empty `ContentMetadata`.
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// A file from the content repository or the location where it can be found.
+#[derive(Debug, Clone)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum FileOrLocation {
+    /// The content of the file.
+    File(Content),
+
+    /// The file is at the given URL.
+    Location(String),
+}
+
+/// The content of a file from the content repository.
+#[derive(Debug, Clone)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Content {
+    /// The content of the file as bytes.
+    pub file: Vec<u8>,
+
+    /// The content type of the file that was previously uploaded.
+    pub content_type: String,
+
+    /// The value of the `Content-Disposition` HTTP header, possibly containing the name of the
+    /// file that was previously uploaded.
+    ///
+    /// See [MDN] for the syntax.
+    ///
+    /// [MDN]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#Syntax
+    pub content_disposition: String,
```

Comment:
I thought both would be required because of [MSC2702](https://github.com/matrix-org/matrix-spec-proposals/pull/2702), but it turns out only `Content-Type` becomes required for servers, so I will change `Content-Disposition` to an `Option`.

And the library we use for `multipart/mixed` always requires a content type for each part, so I thought it would make more sense to always require it. In practice we might need to check what other implementations do to see if it's going to be a problem.

Reply:
Both are optional now.